Add a "Media" WebInspector Element details panel

jernoble opened the pull request:
#### ddfb1a4a5608af37f95f99d837b43db85f48e435
<pre>
Add a &quot;Media&quot; WebInspector Element details panel
<a href="https://bugs.webkit.org/show_bug.cgi?id=265429">https://bugs.webkit.org/show_bug.cgi?id=265429</a>
<a href="https://rdar.apple.com/118865793">rdar://118865793</a>

Reviewed by Devin Rousso and Antoine Quint.

Add a &quot;Media&quot; details panel to the Web Inspector when selecting a &lt;video&gt; or &lt;audio&gt; element.

The contents of the details panel should be broadly similar to the contents of the &quot;Show Media Stats&quot;
panel in desktop WebKit. These contents are collected within HTMLMediaElement::getStatsAsJSON(), put
into a JSON object, and passed to the WebInspector via InspectorDOMAgent. In order to facilitate
crating this JSON stats object, add toJSON/toJSONObject() methods to VideoPlaybackQuality,
AudioTrackConfiguration, and VideoTrackConfiguration. Add a new utility method to create a human
readable string from an extended codec string. And create and populate a new panel within the
WebInspectorUI to display these stats.

* Source/JavaScriptCore/inspector/protocol/DOM.json:
* Source/WTF/wtf/Forward.h:
* Source/WebCore/Modules/mediacontrols/MediaControlsHost.cpp:
(WebCore::MediaControlsHost::sourceType const):
* Source/WebCore/Modules/mediacontrols/MediaControlsHost.h:
* Source/WebCore/Modules/mediasource/VideoPlaybackQuality.cpp:
(WebCore::VideoPlaybackQuality::toJSONObject const):
* Source/WebCore/Modules/mediasource/VideoPlaybackQuality.h:
* Source/WebCore/Modules/webcodecs/VideoColorSpace.cpp: Copied from Source/WebCore/Modules/mediasource/VideoPlaybackQuality.cpp.
(WebCore::VideoColorSpace::toJSON const):
* Source/WebCore/Modules/webcodecs/VideoColorSpace.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::getVideoPlaybackQuality const):
(WebCore::HTMLMediaElement::getStatsAsJSON const):
(WebCore::HTMLMediaElement::sourceType const):
(WebCore::HTMLMediaElement::getVideoPlaybackQuality): Deleted.
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/track/AudioTrackConfiguration.cpp: Copied from Source/WebCore/Modules/mediasource/VideoPlaybackQuality.cpp.
(WebCore::AudioTrackConfiguration::toJSON const):
* Source/WebCore/html/track/AudioTrackConfiguration.h:
* Source/WebCore/html/track/AudioTrackList.cpp:
(WebCore::AudioTrackList::firstEnabled const):
* Source/WebCore/html/track/AudioTrackList.h:
* Source/WebCore/html/track/VideoTrackConfiguration.cpp: Copied from Source/WebCore/Modules/mediasource/VideoPlaybackQuality.cpp.
(WebCore::VideoTrackConfiguration::toJSON const):
* Source/WebCore/html/track/VideoTrackConfiguration.h:
* Source/WebCore/html/track/VideoTrackList.cpp:
(WebCore::VideoTrackList::selectedItem const):
* Source/WebCore/html/track/VideoTrackList.h:
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::getMediaStats):
* Source/WebCore/inspector/agents/InspectorDOMAgent.h:
* Source/WebCore/platform/graphics/CodecUtilities.cpp: Added.
(WebCore::humanReadableStringFromCodecString):
* Source/WebCore/platform/graphics/CodecUtilities.h: Copied from Source/WebCore/Modules/mediasource/VideoPlaybackQuality.cpp.
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
(WI.DOMNode):
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/17.0/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/macOS/14.0/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Views/ElementsTabContentView.js:
(WI.ElementsTabContentView):
* Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.css: Added.
* Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js: Added.
(WI.MediaDetailsSidebarPanel):
(WI.MediaDetailsSidebarPanel.prototype.supportsDOMNode):
(WI.MediaDetailsSidebarPanel.prototype.initialLayout):
(WI.MediaDetailsSidebarPanel.prototype.async layout):
(WI.MediaDetailsSidebarPanel.prototype.get source):
(WI.MediaDetailsSidebarPanel.prototype.set source):
(WI.MediaDetailsSidebarPanel.prototype._updateViewportRow):
(WI.MediaDetailsSidebarPanel.prototype.get viewport):
(WI.MediaDetailsSidebarPanel.prototype.set viewport):
(WI.MediaDetailsSidebarPanel.prototype.get devicePixelRatio):
(WI.MediaDetailsSidebarPanel.prototype.set devicePixelRatio):
(WI.MediaDetailsSidebarPanel.prototype.get quality):
(WI.MediaDetailsSidebarPanel.prototype.set quality):
(WI.MediaDetailsSidebarPanel.prototype.get videoFormat):
(WI.MediaDetailsSidebarPanel.prototype.set videoFormat):
(WI.MediaDetailsSidebarPanel.prototype.get audioFormat):
(WI.MediaDetailsSidebarPanel.prototype.set audioFormat):
(WI.MediaDetailsSidebarPanel.prototype.get video):
(WI.MediaDetailsSidebarPanel.prototype.set video):
(WI.MediaDetailsSidebarPanel.prototype.get audio):
(WI.MediaDetailsSidebarPanel.prototype.set audio):
(WI.MediaDetailsSidebarPanel.prototype.attached):
(WI.MediaDetailsSidebarPanel.prototype.detached):

Canonical link: <a href="https://commits.webkit.org/273777@main">https://commits.webkit.org/273777@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/54b1c1b223c333c1510ef192d808669cf3ab0559

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36615 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15555 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38841 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39276 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32826 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37849 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18029 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12634 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31438 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37176 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13113 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32387 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11485 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11492 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32642 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40520 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/30742 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33177 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32990 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/37432 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/36391 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11750 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9592 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35545 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13430 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/32313 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/43174 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8306 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12167 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/8926 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12635 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
```diff
@@ -3095,4 +3095,19 @@ Protocol::ErrorStringOr<void> InspectorDOMAgent::setAllowEditingUserAgentShadowT
+        return makeUnexpected(errorString);
+
+    if (!element->isMediaElement())
+        return makeUnexpected(errorString);
```
Won't this return an empty string?

jernoble replied:
Ah, it will. Maybe that's fine?

jernoble replied:
Okay, Added an explicit error string in this case.

eric-carlson reviewed (commented):
This looks great to me, but someone more familiar with Web Inspector internals should also take a look

patrickangle commented on Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/17.0/InspectorBackendCommands.js:
```diff
@@ -233,6 +233,7 @@ InspectorBackend.registerCommand("DOM.markUndoableState", ["page"], [], []);
 InspectorBackend.registerCommand("DOM.focus", ["page"], [{"name": "nodeId", "type": "number"}], []);
 InspectorBackend.registerCommand("DOM.setInspectedNode", ["page"], [{"name": "nodeId", "type": "number"}], []);
 InspectorBackend.registerCommand("DOM.setAllowEditingUserAgentShadowTrees", ["page"], [{"name": "allow", "type": "boolean"}], []);
+InspectorBackend.registerCommand("DOM.getMediaStats", null, [{"name": "nodeId", "type": "number"}], ["stats"]);
```
We don't update the compatibility files since these are meant to be a snapshot of what shipped in the version number present in the path. Versions sufficiently beyond this version will use the ToT protocol which is built as part of the stack. (That version number is not determined in open source code).

patrickangle commented on Source/WebInspectorUI/UserInterface/Protocol/Legacy/macOS/14.0/InspectorBackendCommands.js:
```diff
@@ -233,6 +233,7 @@ InspectorBackend.registerCommand("DOM.markUndoableState", ["page"], [], []);
 InspectorBackend.registerCommand("DOM.focus", ["page"], [{"name": "nodeId", "type": "number"}], []);
 InspectorBackend.registerCommand("DOM.setInspectedNode", ["page"], [{"name": "nodeId", "type": "number"}], []);
 InspectorBackend.registerCommand("DOM.setAllowEditingUserAgentShadowTrees", ["page"], [{"name": "allow", "type": "boolean"}], []);
+InspectorBackend.registerCommand("DOM.getMediaStats", null, [{"name": "nodeId", "type": "number"}], ["stats"]);
```
Ditto about not updating versioned legacy protocol files

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.css:
```diff
@@ -0,0 +1,4 @@
+
+.sidebar > .panel.details.media-details > .details-section > .content > .group > .row.simple > .label {
+    min-width: 105px; /* Width of `Historical Figures` or `Optical Size (opsz)` in English. */
```
Can we update this comment and use a width based on the longest English string in the Media panel?

patrickangle commented on Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
```diff
@@ -1321,6 +1321,19 @@ WI.DOMNode = class DOMNode extends WI.Object
+    {
+        return new Promise((resolve, reject) => {
+            let target = WI.assumingMainTarget();
+            target.DOMAgent.getMediaStats(this.id, (error, stats) => {
```
Protocol commands should be able to be called asynchronously by not passing a callback. So this entire function should be able to be written as:
```
async getMediaStats()
{
    let target = WI.assumingMainTarget();
    return await target.DOMAgent.getMediaStats(this.id);
}
```

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,240 @@
+            return;
+
+        this._source = source;
+        this._sourceRow.value = WI.UIString(source);
```
This isn't a valid way to express a UIString in Web Inspector. The localized string file is generated using the `update-webkit-localizable-strings`, and this will not produce valid results. Localized strings should be thought of as constants. For something like this you will want to have a switch-case that does the conversion from protocol values to a localized string.

patrickangle commented on Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
```diff
@@ -2000,3 +2000,33 @@ localizedStrings["unsupported version"] = "unsupported version";
+localizedStrings["Audio Details"] = "Audio Details";
+localizedStrings["Transfer Function"] = "Transfer Function";
+localizedStrings["Color Primaries"] = "Color Primaries";
+localizedStrings["Matrix Coefficients"] = "Matrix Coefficients";
```
Many, if not all, of these strings really need unique keys and descriptions since they are very domain-specific, and thus that context is critical for localizers. e.g. `WI.UIString("Matrix Coefficients", "Matrix Coefficients @ Media Details Sidebar Panel", "Title label for the matrix coefficients of a video/audio element")`

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,240 @@
+        this.element.appendChild(this._audioSection.element);
+    }
+
+    async layout()
```
Making `layout` async is likely to be unpredictable, since this means that children views (which admittedly you don't have any here) would get laid out before this view has finished laying out.

Generally layout is not when we should be doing the work of getting the underlying data from the backend. This could be particularly costly during resize, as `layout` will be called once per frame during a resize.

One possible solution would be to perform the asynchronous work in a private function you invoke with your update timer and then call `needsLayout` once you have updated info available.

(Also keep in mind that calls to the backend are not instant, particularly when inspecting content on another device via Remote Inspection.)

jernoble replied:
Frankly, `needsLayout()` and `layout()` seem to be entirely mis-used in a lot of the Web Inspector code, since the implementations are about updating UI from the model, and not about layout itself. But sure, I'll refactor all this to move the pushing of new values into the UI outside `layout()`.

patrickangle replied:
To be clear, my concern is that we shouldn’t be doing async work inside layout to ask the frontend’s model to be updated from the backend, I don’t see an issue with updating from the frontend’s current representation of the model here.

dcrousso replied:
the pattern we use in Web Inspector is that `layout()` is the place for doing (major) DOM modifications (and ideally that's all it should do)

as @patrickangle said, we want to avoid doing other (potentially expensive) things in `layout()` because it can get called a _lot_ (i.e. resizing the window, other parts of the UI moving around, some child `WI.View` changes that dirties the parent, etc.)

the issue here is just that you could be calling `getMediaStats()` in response to any of those other scenarios in addition to when you're calling `needsLayout()`

i think since youre doing a "fetch every 250ms regardless of everything else happening", this actually probably should be some sort of combination of "fetch the data from the backend, and once we have it we `needsLayout()`, and then in `layout()` we update the UI and start the process again" but making sure that we never have more than one `DOM.getMediaStats` in-flight at once

jernoble replied:
Right, but updating the DOM from the model isn't _layout_. Perhaps updating the DOM from the model should _trigger layout_, but it isn't layout in and of itself.

jernoble replied:
Regardless, the async updating of information was moved out of the `layout()` method (which no longer exists because it's not needed; CSS is sufficient here).

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,240 @@
+        this.audio = stats.audio;
+    }
+
+    get source() {
```
All of these getter/setters appear to be only used within this class, so should be prefixed with an underscore and moved below to the `// Private` section.

jernoble replied:
Is there no convention to use actual private methods and members in WI? (E.g., `#source` rather than `_source`)

patrickangle replied:
There isn't currently because until recently there was no way in Web Inspector to see those properties or their values, so we previously didn’t use them because debugging wasn’t practical. That changed earlier this year, so IMO we could indeed start using private members and functions.

dcrousso replied:
the debugging story with private fields isn't super great either, as there's not a good way to reach them via the Console (yet).  i think there may also be some (unresolved) performance impact when using private fields :/

also, any of the `get foo` that aren't used should probably be removed

jernoble replied:
None of those seem like good reasons to avoid private fields; if making accessors private is important, they need to be actually private, as anything else is just convention. 

It's super weird to have write-only properties; if `get foo` should be removed, then all `set foo` accessors should be turned into `setFoo(foo)` functions.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,240 @@
+        super.attached();
+        if (this._updateTimer)
+            clearInterval(this._updateTimer);
+        this._updateTimer = setInterval(() => this.needsLayout(), 250);
```
It isn't generally a good idea to poll like this for updates. The backend should be sending an event over the protocol with updates when they exist instead of the frontend having to continually ask the backend. Additionally, a fixed interval doesn't take into account it may take longer than that interval to perform the update, at which point you may have multiple updates in flight at once, or worse you may have a growing backlog of update requests still waiting on the backend.

jernoble replied:
There's no event for the most frequent updated value in these stats (the dropped frame count). It's a poll-or-nothing here.

patrickangle replied:
I see. Polling it is, I guess. I think we have WI.Throttler that should make it easier to set this interval and address some of the issues I mentioned, but it might not be async-ready currently

dcrousso replied:
while it's definitely not ideal to use frontend polling, if the backend has no way of knowing when one of the values has changed then yeah i suppose we have no other choice

given that, the most important thing we should ensure is that we never have more than one command in-flight at the same time

as such, id probably do something like this
```js
layout()
{
    if (this._updatedMediaStats) {
        this._updatedMediaStats = false;

        // modify relevant DOM stuff from member variable(s) saved below

        this._getMediaStats();
    }
}

attached()
{
    super.attached();

    this._getMediaStats();
}

detached()
{
    this._getMediaStatsPromise = null;
    this._getMediaStatsTimestamp = 0;
    if (this._getMediaStatsTimeoutIdentifier)
        clearTimeout(this._getMediaStatsTimeoutIdentifier);
        this._getMediaStatsTimeoutIdentifier = undefined;
    }

    super.detached();
}

_getMediaStats()
{
    if (this._getMediaStatsPromise || this._getMediaStatsTimeoutIdentifier)
        return;

    const delay = 250;
    let timestamp = Date.now();
    let timeout = delay - timestamp - this._getMediaStatsTimestamp;
    if (timeout > 0) {
        this._getMediaStatsTimeoutIdentifier = setTimeout(() => {
            this._getMediaStatsTimeoutIdentifier = undefined;
            this._getMediaStats();
        }, timeout); 
        return;
    }

    this._getMediaStatsTimestamp = timestamp;
    this._getMediaStatsPromise = this.domNode.getMediaStats();
    this._getMediaStatsPromise.then((mediaStats) => {
        if (!this._getMediaStatsPromise)
            return;

        this._getMediaStatsPromise = null;

        // process and save `mediaStats` to member variable(s)

        this._updatedMediaStats = true;
        this.needsLayout();
    });
}
```
which would ensure that we
- don't have more than one `DOM.getMediaStats` in-flight at the same time
- don't send a second `DOM.getMediaStats` until we've actually displayed the results of the first
- don't modify the DOM outside of `layout()`

ideally we'd use a `Throttler` as @patrickangle suggested, but it (currently) doesnt support the `callback` being `async` and as such wouldn't be able to ensure that we dont send a second `DOM.getMediaStats` until the first has finished

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.css:
```diff
@@ -0,0 +1,4 @@
+
```
This file needs a license header as well. The other CSS files will have one formatted for CSS (since `//` isn't valid in CSS).

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,240 @@
+    {
+        super.initialLayout();
+
+        this._sourceRow = new WI.DetailsSectionSimpleRow(WI.UIString("Source"));
```
All private members should be defined as existing in the constructor to avoid churn on the prototype of objects e.g. make sure all the properties of the object is know at creation time with `this._sourceRow = null`, etc.. It's also just nice to be able to reason about all the properties of an object from the constructor, even the "private" ones.

Web Inspector code isn't perfect about doing this, but it is the standard for new code being written last time Inspector folks discussed it.

(@dcrousso always did a better job explaining the technical reason for this than I do)

dcrousso replied:
the idea is to try to group JS object structure transitions as much as possible

i think it would be ideal to at least declare (but maybe not initialize) all members in the `constructor`, but i see a bunch of cases where we declare UI related things in `initialLayout` since it only runs once, so one could argue precedent 😅

to be clear, if you want to do the more ideal thing it'd be something like this
```js
constructor()
{
    this._sourceRow = null;
}

initialLayout()
{
    this._sourceRow = ...
}
```

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,240 @@
+    }
+
+    get source() {
+        return this._source;
```
Also ditto on making sure there is a `this._source = null;` in the constructor to make it part of the class prototype at creation. (Same with all the other getter/setter pairs and the update timer and any other properties you set I may have missed)

patrickangle commented on Source/JavaScriptCore/inspector/protocol/DOM.json:
```diff
@@ -695,6 +695,16 @@
+                { "name": "nodeId", "$ref": "NodeId", "description": "Id of the node to retrieve mediastats for." }
+            ],
+            "returns": [
+                { "name": "attributes", "type": "object", "description": "An interleaved array of node attribute names and values." }
```
A raw object doesn't make for a good protocol value because it isn't possible from the definition of the protocol to know even the general shape of this object without more information. We also won't be able to have future differences appear in the legacy protocol versions, since the contents of the attributes is hidden behind the opaque object.

For an example of a somewhat similar object that we have that is a collection of attributes and how the protocol type is used, see `ContextAttributes` in Canvas.json:42.

jernoble replied:
JavaScript objects are inherently introspectible. The entire shape, much less the general one, can be inferred from the result.  So the justification for an explicit description of the type passed across the protocol boundary must be different than this. 

I also don't see anywhere within the Legacy protocol definitions where the types have changed across protocol versions. Can you point one out?

patrickangle replied:
I’ll dig up some when I’m at a computer tomorrow morning.

dcrousso replied:
> JavaScript objects are inherently introspectible.

it's not always the case that the frontend is using JavaScript (e.g. ITML inspection)

> The entire shape, much less the general one, can be inferred from the result. So the justification for an explicit description of the type passed across the protocol boundary must be different than this.

since the shape is not specified in order for a frontend to be "safe" it would have to have _significantly_ more complex logic around "what properties exist" and "what is the type of this property", potentially handling any possible variation (e.g. what if `foo` is a number vs an array vs an object vs etc.?)

generally speaking the entire reason the protocol exists is to allow the frontend and backend to guarantee a common understanding of what each is capable of and the shape of the data being passed between them.  sure, we technically could just use `"type": "any"` everywhere, but then what's the point of having the protocol in the first place?

> I also don't see anywhere within the Legacy protocol definitions where the types have changed across protocol versions. Can you point one out?

`CSS.CSSRule` added a `isImplicitlyNested` property in iOS17.0/macOS14.0 that didn't exist in iOS16.4/macOS13.3

dcrousso commented on Source/WebCore/platform/graphics/CodecUtilities.cpp:
```diff
@@ -0,0 +1,163 @@
+        StringBuilder builder;
+        builder.append("AVC (");
+        switch (parameters->profileIDC) {
+        case 66: builder.append("Baseline Profile, "); break;
```
do we expect these strings to appear in Web Inspector?  if so, i think we need to convert these into a protocol enum and have the frontend handle displaying the text so that it can be localized

jernoble replied:
No, I'd much, much rather require them to be localized in the WebContent process. There's already too many places where localized strings need to be added.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FontDetailsSidebarPanel.js:
```diff
@@ -29,4 +29,11 @@ WI.FontDetailsSidebarPanel = class FontDetailsSidebarPanel extends WI.GeneralSty
+
+    supportsDOMNode(nodeToInspect)
+    {
+        if (nodeToInspect.isMediaElement())
```
id rather this be a followup since it doesn't really have anything to do with adding a new Media panel

we also probably should disable this for a whole bunch of other kinds of elements too

jernoble replied:
This was requested because the "number of tabs" in the Element sidebar was getting too large.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,240 @@
+    constructor(delegate)
+    {
+        super("media-details", WI.UIString("Media"));
+        this.element.classList.add("media-details");
```
the `identifier` provided to `super` should already be added as a CSS class, so i think this is redundant

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,240 @@
+        this._audioFormatRow = new WI.DetailsSectionSimpleRow(WI.UIString("Audio Format"));
+
+        let generalGroup = new WI.DetailsSectionGroup([this._sourceRow, this._viewportRow, this._framesRow, this._resolutionRow, this._videoFormatRow, this._audioFormatRow
+                                                       ]);
```
Style: this should be on the same line

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,240 @@
+        return this._source;
+    }
+
+    set source(source) {
```
Style: the `{` should be on its own line

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,240 @@
+
+    set viewport(viewport) {
+        if (this._viewport?.width === viewport?.width
+            && this._viewport?.height === viewport?.height)
```
id suggest using `Object.shallowEqual` for these kinds of comparisons instead

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,240 @@
+        this._videoCodecRow.value = video?.codec;
+        this._transferRow.value = video?.colorSpace?.transfer;
+        this._primariesRow.value = video?.colorSpace?.primaries;
+        this._matrixRow.value = video?.colorSpace?.matrix;
```
does these need a fallback values?

jernoble replied:
No, that's a job for the value->UI conversion to determine the correct thing to display for `null` or `undefined` values.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,240 @@
+
+        this._audio = audio;
+        this._audioSection.element.classList.toggle("hidden", !audio);
+        this._audioCodecRow.value = audio?.codec;
```
ditto

dcrousso reviewed (changes requested):
overall this is pretty awesome!  i think we just need some structural adjustments to the protocol and UI bits

graouts commented on Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
```diff
@@ -3095,4 +3106,84 @@ Protocol::ErrorStringOr<void> InspectorDOMAgent::setAllowEditingUserAgentShadowT
+{
+    Protocol::ErrorString errorString;
+
+    Element* element = assertElement(errorString, nodeId);
```
Could use `auto*` presumably.

graouts commented on Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
```diff
@@ -3095,4 +3106,84 @@ Protocol::ErrorStringOr<void> InspectorDOMAgent::setAllowEditingUserAgentShadowT
+    if (!element)
+        return makeUnexpected(errorString);
+
+    if (!element->isMediaElement())
```
Could use `dynamicDowncast<HTMLMediaElement>` on `element` and have an early return if the value is `nullptr` here. That way you can combine the `isMediaElement()` test here and the `downcast<>` below.

graouts commented on Source/WebCore/platform/graphics/CodecUtilities.cpp:
```diff
@@ -0,0 +1,257 @@
+            return WEB_UI_STRING_KEY("AAC-LC", "AAC-LC Codec String", "Codec Strings");
+        case 34:
+            WEB_UI_STRING_KEY("MP3", "MP3 Codec String", "Codec Strings");
+        }
```
No handling of the default value `0`?

jernoble replied:
No, it should fall through to "MP4 Audio".

graouts commented on Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
```diff
@@ -1321,6 +1321,12 @@ WI.DOMNode = class DOMNode extends WI.Object
             this.showLayoutOverlay();
     }
+
+    async getMediaStats()
```
I was going to suggest using a property here, since that would be more aligned with the Inspector coding style, but it appears you can't have `async get mediaStats()`. You could still do that by returning a `Promise`. It's really up to you.

graouts commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,319 @@
+
+    #cancelUpdateTimer()
+    {
+        if (this.#updateTimer)
+            clearTimeout(this.#updateTimer);
```
I _think_ it's safe to call `clearTimeout` with a null value, it'll just not do anything.

graouts reviewed (approved):
A general comment would be to use `const` instead of `let` for JS values that aren't going to be re-assigned, but it's a very "me" thing.

dcrousso commented on Source/WebCore/platform/graphics/CodecUtilities.cpp:
```diff
@@ -0,0 +1,257 @@
+            builder.append(WEB_UI_STRING_KEY("VP8", "VP8 (Codec String)", "Codec Strings"));
+        else if (parameters->codecName == "vp09"_s)
+            builder.append(WEB_UI_STRING_KEY("VP9", "VP9 (Codec String)", "Codec Strings"));
+        builder.append(" (");
```
this will have an empty space at the start if the `codecName` is not `"vp08"` or `"vp09"` is that expected/possible?

jernoble replied:
It's not currently possible for that value to be anything other than "vp08" or "vp09" and it's very unlikely for that to change.

dcrousso commented on Source/WebCore/platform/graphics/CodecUtilities.cpp:
```diff
@@ -0,0 +1,257 @@
+        case 2:
+            return WEB_UI_STRING_KEY("AAC-LC", "AAC-LC Codec String", "Codec Strings");
+        case 34:
+            WEB_UI_STRING_KEY("MP3", "MP3 Codec String", "Codec Strings");
```
should this be a `return`?

jernoble replied:
omg, yes.

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
```diff
@@ -1321,6 +1321,12 @@ WI.DOMNode = class DOMNode extends WI.Object
+    async getMediaStats()
+    {
+        let target = WI.assumingMainTarget();
+        return (await target.DOMAgent.getMediaStats(this.id)).mediaStats;
```
Style: we prefer putting the `await` into its own variable so that it's easier to read
```js
let {mediaStats} = await target.DOMAgent.getMediaStats(this.id);
return mediaStats;
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,319 @@
+WI.MediaDetailsSidebarPanel = class MediaDetailsSidebarPanel extends WI.DOMDetailsSidebarPanel
+{
+    // Static
+    static StyleClassName = "media";
```
Style: missing space after comment

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,319 @@
+    static StyleClassName = "media";
+
+    // Public
+    constructor(delegate)
```
ditto

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,319 @@
+    // Public
+    constructor(delegate)
+    {
+        class MediaDetailsSectionSimpleRow extends WI.DetailsSectionSimpleRow
```
i dont think we've ever created an inline `class` like this.  since it doesnt actually capture anything i'd much prefer it be moved outside of `class MediaDetailsSidebarPanel` so that we dont have to create it more than once
```js
WI.MediaDetailsSidebarPanel.Row = class MediaDetailsSidebarPanelRow
{
    ...
};
```

jernoble replied:
This is done this way because I can't guarantee that `WI.DetailsSectionSimpleRow` has been evaluated at the time in which this script is parsed. It throws an invalid type error if I do it at file scope.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,319 @@
+    {
+        class MediaDetailsSectionSimpleRow extends WI.DetailsSectionSimpleRow
+        {
+            set pendingValue(value)
```
Style: please add a `// Public` before these two methods

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,319 @@
+
+        super("media-details", WI.UIString("Media"));
+
+        this.#ui.sourceRow = new MediaDetailsSectionSimpleRow(WI.UIString("Source", "Source @ Media Sidebar", "Title for Source row in Media Sidebar"));
```
we dont want to actually create these object inside the `constructor` because it gets called even if the the details sidebar panel is never selected (i.e. no need to create all these `MediaDetailsSectionSimpleRow` until we actually show the details sidebar panel for the first time).  as such, id ask for the creation of the properties to be kept in the `constructor` (or via a class field) but the initialization of the value to be moved to `initialLayout`, which i think should just really end up being moving all this to `initialLayout` since you already have class fields for the properties :)

jernoble replied:
This would mean adding null-checks everywhere because I can't guarantee that initialLayout() has been called before attached(). This is (IMO) an example of a premature optimization. The cost of creating these rows is very low, and in addition to being the view, each row also stores its displayed value, so they pull double duty.

I did originally have all this in initialLayout(), and moved them into the constructor for this very reason.

dcrousso replied:
perhaps in this case it might be premature because these objects are small, but it is the general practice of Web Inspector to not initialize view related stuff inside the `constructor` instead of `initialLayout` precisely because we've had issues with that in the past (e.g. i spent an entire summer moving things around because the aggregate of many cases like this got so bad)

while null checks would be fine you also could use `this.didInitialLayout` as is done in many other places.  in fact i think you would only need to wrap the call to `this.#startUpdateTimer();` inside `attached()` since that's really the only entrypoint that isn't during/after `initialLayout()`

jernoble replied:
If there's a problem with creating all the panels before they're needed, then that needs to be resolved by... not creating all the panels until they're needed, and not by micro-optimizing the constructors of each individual panel.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,319 @@
+        return nodeToInspect.isMediaElement();
+    }
+
+    attached()
```
these should be moved to the `// Protected` section since they're more general overrides than just from `DOMDetailsSidebarPanel`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,319 @@
+
+    // Private
+
+    #values = {
```
im not sure how i feel about having subobjects like this.  on the one hand, i do appreciate how it organizes values vs elements.  on the other hand, it means every time we need to get a value we have to jump across two objects instead of just one (i.e. `this.#foo` vs `this.#values.foo`/`this.#ui.foo`).

id prefer keeping things flat as it allows for future developers to more quickly see the state of `WI.MediaDetailsSidebarPanel` if it's logged in inspector^2, especially since all of the UI properties are already suffixed with `Row`, which also has the added benefit of some nice parity (i.e. having `source` and `sourceRow` be side-by-side makes it really easy to connect model and UI)

jernoble replied:
I disagree. I think they're much more comprehensible as they are.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,319 @@
+        videoFormat: null,
+        audioFormat: null,
+        video: null,
+        audio: null
```
Style: missing `,`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,319 @@
+        audioCodecRow: null,
+        sampleRateRow: null,
+        channelsRow: null,
+        audioSection: null
```
ditto

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,319 @@
+        this.needsLayout();
+    }
+
+    #updateViewportRow()
```
NIT: id inline this function since it's only called once

jernoble replied:
Aha, this is actually a bug. This method should be called from #setDevicePixelRatio as well.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,319 @@
+        if (this.#values.video === video)
+            return;
+
+        if (this.#values.video?.width === video?.width
```
are you doing this because `Object.shallowEqual` doesn't support subobjects?  i cant think of a reason why we wouldn't want to support that so id suggest using `Object.shallowEqual` here and modifying the definition of `Object.shallowEqual` like so
```diff
-if (aValue !== bValue && !Array.shallowEqual(aValue, bValue))
+if (aValue !== bValue && !Object.shallowEqual(aValue, bValue))
```
(and add a test or two in `LayoutTests/inspector/unit-tests/object-utilities.html`)

jernoble replied:
Well, if we modify the shallowEqual method to no longer be shallow... It should probably not be called shallowEqual either. ;)

dcrousso replied:
we definitely should update the name too :)

i dont mind this being done in a followup.  i just think it's a good idea overall

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,319 @@
+            return;
+
+        this.#values.video = video;
+        this.#ui.videoSection.element.classList.toggle("hidden", !video);
+        this.#ui.resolutionRow.pendingValue = WI.UIString("%dx%d (%dfps)").format(video?.width ?? 0, video?.height ?? 0, video?.framerate ?? 0);
+        this.#ui.resolutionRow.element.classList.toggle("hidden", !video);
+        this.#ui.videoCodecRow.pendingValue = video?.codec;
```
im pretty sure `WI.DetailsSection` should support automatically hiding itself if none of the contained `WI.DetailsSectionSimpleRow` have any value.  is that not the case here?  perhaps it's because you're always populating certain rows even if there's no real value to give it?  if `WI.DetailsSection` does behave that way id ask you to remove these manual `"hidden"` and instead just prevent values from always being set (e.g. use `""` as a fallback if the relevant data isn't available)

along those lines, what does the Media details sidebar panel show when an empty `<video>` is selected in the Elements Tab?  what about for a `<video src="...">` that hasn't played yet?  i trust you to know what _should_ be shown, so long as there is at least something shown (i.e. i would _not_ want it to be an empty details sidebar panel with no placeholder text of any kind)

jernoble replied:
Good point, "Source" should at least read "Empty". But yes, it'd be mostly empty because there are no video or audio tracks yet.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,319 @@
+        this.#ui.videoSection.element.classList.toggle("hidden", !video);
+        this.#ui.resolutionRow.pendingValue = WI.UIString("%dx%d (%dfps)").format(video?.width ?? 0, video?.height ?? 0, video?.framerate ?? 0);
+        this.#ui.resolutionRow.element.classList.toggle("hidden", !video);
+        this.#ui.videoCodecRow.pendingValue = video?.codec;
+        this.#ui.transferRow.pendingValue = video?.colorSpace?.transfer;
+        this.#ui.primariesRow.pendingValue = video?.colorSpace?.primaries;
+        this.#ui.matrixRow.pendingValue = video?.colorSpace?.matrix;
```
it's really unfortunate that we keep having to optional chain against the same value over and over.  is there any way we can refactor this to have a single `if (video) {` check at the top (and same for `if (video.colorSpace) {`) so that we can avoid doing all these checks over and over and over?

jernoble replied:
No, because then we'd need `} else {` blocks for each, so they clear out the values when going from having `video` to not, and etc. This is much cleaner.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,319 @@
+
+    #setAudio(audio)
+    {
+        if (this.#values.audio === audio)
```
this is already checked by `Object.shallowEqual`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -0,0 +1,319 @@
+            this.#ui.channelsRow.pendingValue = WI.UIString("Stereo");
+            break;
+        default:
+            this.#ui.channelsRow.pendingValue = audio?.numberOfChannels ?? WI.UIString("Unknown");
```
rather than show "Unknown" would it be better to just hide the row entirely?

dcrousso commented on LayoutTests/inspector/dom/getMediaStats.html:
```diff
@@ -0,0 +1,78 @@
+    suite.addTestCase({
+        name: "GetMediaStats-InvalidElement",
+        description: "Call GetMediaStats on an invalid element.",
+        test: async () => {
```
NIT: you can just write `async test() {`

ditto below

dcrousso commented on LayoutTests/inspector/dom/getMediaStats.html:
```diff
@@ -0,0 +1,78 @@
+            let nodeId = await documentNode.querySelector("#notAVideo");
+            let domNode = WI.domManager.nodeForId(nodeId);
+            InspectorTest.assert(domNode, "DOMNode exists");
+            try {
```
id suggest using `InspectorTest.expectException` instead as it also logs the error (which we can use to verify that it's throwing at the right place in `InspectorDOMAgent::getMediaStats`
```js
await InspectorTest.expectEqual(() => {
    await domNode.getMediaStats();
});
```

dcrousso commented on LayoutTests/inspector/dom/getMediaStats.html:
```diff
@@ -0,0 +1,78 @@
+            let nodeId = await documentNode.querySelector("#emptyVideo");
+            let domNode = WI.domManager.nodeForId(nodeId);
+            InspectorTest.assert(domNode, "DOMNode exists");
+            try {
```
id recommend not including a `try` at all if we're not expecting an exception as that way if it does happen the test harness will print out a whole bunch of relevant info to fix it

ditto below

dcrousso commented on LayoutTests/inspector/dom/getMediaStats.html:
```diff
@@ -0,0 +1,78 @@
+                let stats = await domNode.getMediaStats();
+                InspectorTest.log("PASS: Did not throw when calling getMediaStats on video#emptyVideo");
+                InspectorTest.expectEqual(stats.video, undefined, "Empty Video elements should have empty stats.video");
+                InspectorTest.expectEqual(stats.audio, undefined, "Empty Video elements should have empty stats.audio");
```
NIT: id take this a step further and check that the property doesnt even exist since we ideally dont even want to send it if it's not set
```js
InspectorTest.expectFalse("video" in stats, "Empty Video elements should not provide a stats.video".)
```

dcrousso commented on Source/WebCore/html/track/AudioTrackConfiguration.h:
```diff
@@ -54,6 +54,8 @@ class AudioTrackConfiguration : public RefCounted<AudioTrackConfiguration> {
     uint64_t bitrate() const { return m_state.bitrate; }
     void setBitrate(uint64_t bitrate) { m_state.bitrate = bitrate; }
 
+    Ref<JSON::Object> toJSON() const;
```
NIT: do we still need these C++ `toJSON` now that we're using Web Inspector protocol types?

jernoble replied:
Actually no; this could be part of the follow-up patch that brings the same UI/properties to "Show Media Stats" (they'll be used there).

dcrousso reviewed (approved):
overall this looks great!  nice work!  :)

i think there's only a couple things that definitely need to be changed, as well as some other Style/NIT

i'm still not really sold on the idea of using private fields since they're not as performant and also dont really have a good debugging story in Web Inspector (yet), but i do think we should eventually move in that direction so i suppose there's no time like the present to give it a shot

> A general comment would be to use `const` instead of `let` for JS values that aren't going to be re-assigned, but it's a very "me" thing.

Web Inspector style is to only use `const` for things that are identical across Web Inspector sessions (i.e. true constants) (e.g. numbers, literal strings, objects/arrays/etc. that aren't modified after initial creation and are built using other constant data, etc.).  put another way, we use `const` sorta like how you can use `constexpr` in C++

jernoble committed:
Add a "Media" WebInspector Element details panel
https://bugs.webkit.org/show_bug.cgi?id=265429
rdar://118865793

Reviewed by Devin Rousso and Antoine Quint.

Add a "Media" details panel to the Web Inspector when selecting a <video> or <audio> element.

The contents of the details panel should be broadly similar to the contents of the "Show Media Stats"
panel in desktop WebKit. These contents are collected within HTMLMediaElement::getStatsAsJSON(), put
into a JSON object, and passed to the WebInspector via InspectorDOMAgent. In order to facilitate
crating this JSON stats object, add toJSON/toJSONObject() methods to VideoPlaybackQuality,
AudioTrackConfiguration, and VideoTrackConfiguration. Add a new utility method to create a human
readable string from an extended codec string. And create and populate a new panel within the
WebInspectorUI to display these stats.

* Source/JavaScriptCore/inspector/protocol/DOM.json:
* Source/WTF/wtf/Forward.h:
* Source/WebCore/Modules/mediacontrols/MediaControlsHost.cpp:
(WebCore::MediaControlsHost::sourceType const):
* Source/WebCore/Modules/mediacontrols/MediaControlsHost.h:
* Source/WebCore/Modules/mediasource/VideoPlaybackQuality.cpp:
(WebCore::VideoPlaybackQuality::toJSONObject const):
* Source/WebCore/Modules/mediasource/VideoPlaybackQuality.h:
* Source/WebCore/Modules/webcodecs/VideoColorSpace.cpp: Copied from Source/WebCore/Modules/mediasource/VideoPlaybackQuality.cpp.
(WebCore::VideoColorSpace::toJSON const):
* Source/WebCore/Modules/webcodecs/VideoColorSpace.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::getVideoPlaybackQuality const):
(WebCore::HTMLMediaElement::getStatsAsJSON const):
(WebCore::HTMLMediaElement::sourceType const):
(WebCore::HTMLMediaElement::getVideoPlaybackQuality): Deleted.
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/track/AudioTrackConfiguration.cpp: Copied from Source/WebCore/Modules/mediasource/VideoPlaybackQuality.cpp.
(WebCore::AudioTrackConfiguration::toJSON const):
* Source/WebCore/html/track/AudioTrackConfiguration.h:
* Source/WebCore/html/track/AudioTrackList.cpp:
(WebCore::AudioTrackList::firstEnabled const):
* Source/WebCore/html/track/AudioTrackList.h:
* Source/WebCore/html/track/VideoTrackConfiguration.cpp: Copied from Source/WebCore/Modules/mediasource/VideoPlaybackQuality.cpp.
(WebCore::VideoTrackConfiguration::toJSON const):
* Source/WebCore/html/track/VideoTrackConfiguration.h:
* Source/WebCore/html/track/VideoTrackList.cpp:
(WebCore::VideoTrackList::selectedItem const):
* Source/WebCore/html/track/VideoTrackList.h:
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::getMediaStats):
* Source/WebCore/inspector/agents/InspectorDOMAgent.h:
* Source/WebCore/platform/graphics/CodecUtilities.cpp: Added.
(WebCore::humanReadableStringFromCodecString):
* Source/WebCore/platform/graphics/CodecUtilities.h: Copied from Source/WebCore/Modules/mediasource/VideoPlaybackQuality.cpp.
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
(WI.DOMNode):
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/17.0/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/macOS/14.0/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Views/ElementsTabContentView.js:
(WI.ElementsTabContentView):
* Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.css: Added.
* Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js: Added.
(WI.MediaDetailsSidebarPanel):
(WI.MediaDetailsSidebarPanel.prototype.supportsDOMNode):
(WI.MediaDetailsSidebarPanel.prototype.initialLayout):
(WI.MediaDetailsSidebarPanel.prototype.async layout):
(WI.MediaDetailsSidebarPanel.prototype.get source):
(WI.MediaDetailsSidebarPanel.prototype.set source):
(WI.MediaDetailsSidebarPanel.prototype._updateViewportRow):
(WI.MediaDetailsSidebarPanel.prototype.get viewport):
(WI.MediaDetailsSidebarPanel.prototype.set viewport):
(WI.MediaDetailsSidebarPanel.prototype.get devicePixelRatio):
(WI.MediaDetailsSidebarPanel.prototype.set devicePixelRatio):
(WI.MediaDetailsSidebarPanel.prototype.get quality):
(WI.MediaDetailsSidebarPanel.prototype.set quality):
(WI.MediaDetailsSidebarPanel.prototype.get videoFormat):
(WI.MediaDetailsSidebarPanel.prototype.set videoFormat):
(WI.MediaDetailsSidebarPanel.prototype.get audioFormat):
(WI.MediaDetailsSidebarPanel.prototype.set audioFormat):
(WI.MediaDetailsSidebarPanel.prototype.get video):
(WI.MediaDetailsSidebarPanel.prototype.set video):
(WI.MediaDetailsSidebarPanel.prototype.get audio):
(WI.MediaDetailsSidebarPanel.prototype.set audio):
(WI.MediaDetailsSidebarPanel.prototype.attached):
(WI.MediaDetailsSidebarPanel.prototype.detached):

Canonical link: https://commits.webkit.org/273777@main
